index.bs

```diff
@@ -1804,6 +1804,96 @@ The [=remote end steps=] with |session| and <var ignore>command parameters</var>
 
 </div>
 
+#### The browser.createPartition Command #### {#command-browser-createPartition}
+
+The <dfn export for=commands>browser.createPartition</dfn> command creates a storage partition for a group of browsing contexts.
```

Comment:
We should absolutely cross-reference "storage partition".

That said, I don't think a storage partition is what we want, as I believe I said in the F2F, and as https://github.com/w3c/webdriver-bidi/issues/289#issuecomment-1721026858 kinda goes into.

We want something that encompasses the global partition (as is the case of ephemeral sessions and the like), whereas specs mostly use "storage partition" to refer to (e.g.) the storage addressable from example.com as the top-level site.

Based on that discussion, we want something more like "user agent" (https://infra.spec.whatwg.org/#user-agent), though the terminology does seem somewhat clumsy here, especially given how we already use "user agent" in WebDriver (Classic/BiDi).

That said, there is some variation between implementations (e.g., whether they share DNS cache results), which we may or may not want to overly specify in WebDriver.

cc @jyasskin, given I was speaking to him at TPAC about this

Reply:
Yep, see https://github.com/privacycg/storage-partitioning and https://w3ctag.github.io/privacy-principles/#dfn-partition for definitions of "partition". [Browsing contexts](https://html.spec.whatwg.org/multipage/document-sequences.html#browsing-context) can navigate between partitions, and HTML is [recommending](https://html.spec.whatwg.org/multipage/document-sequences.html#infrastructure-for-sequences-of-documents) that specs should usually use [navigables](https://html.spec.whatwg.org/multipage/document-sequences.html#navigable), [traversibles](https://html.spec.whatwg.org/multipage/document-sequences.html#traversable-navigables), or [top-level traversibles](https://html.spec.whatwg.org/multipage/document-sequences.html#top-level-traversables) instead of browsing contexts.

With the "user agent" definition in Infra, I was trying to capture the instances of an implementation that are almost completely isolated from each other, so that existing specs that use "A user agent has a quuxinator map" correctly separate those maps. Searching in https://w3c.github.io/webdriver/, I don't see uses of "user agent" that conflict with the Infra definition, although I could be missing them, and some might be more precise as "implementation". For this command, I think the definition might be that it "creates a new [user agent](https://infra.spec.whatwg.org/#user-agent) for the same [implementation](https://infra.spec.whatwg.org/#implementation) as the current" ... "browser process"? "session"? ... ", and returns a new session inside that user agent"?

The current text isn't very clear about what implicit context there is or how you target requests to particular UAs/partitions/sessions. It seems like there's both an implicit browser process and an implicit session, and you can also create new explicit sessions, but nothing uses the `sessionId` returned when you create a new session... Adding profiles (/UAs/partitions) will need the other creation commands to be able to make their thing inside a particular one of those.

Reply:
In this proposal, there is no session created for the partition: all partitions are managed by the same session. The advantage is that the same session can create a fresh state for its operations. In practical terms, we want to have something that would allows to implement https://pptr.dev/api/puppeteer.browsercontext/#example In Chrome, that corresponds to an incognito window (but you can have multiple for automation).

How exactly to name this is an open question.

Reply:
I think specifying it as a creation of a new user agent might be what we need but at time same time we do not want to have a new session and a completely fresh browser process for this feature (if that is the case, I think it would be easier to start a new session instead). Thanks for the links and ideas, I will read up on the specs you linked and see how I can rephrase. And sorry if I missed the previous links to the discussion about it. 

> Adding profiles (/UAs/partitions) will need the other creation commands to be able to make their thing inside a particular one of those.

The current draft modifies the browsing context creation command to accept which partition the context should be created in. Afterwards, the session can address that context by using the context ID and can also determine which contexts are in which partitions based on the partition ID in `browsingContext.Info`.

Reply:
You have a good point that WebDriver needs to be more precise about what it means by "user agent" and "implementation" than most web specs. After all, WebDriver involves at least 2 different "user agents" and "implementations": the software that drives web browsers, and the web browsers themselves. Infra's definition is compatible with this: the key part of my change to that definition is that a single implementation can be configured to act as more than one UA. The application to browser profiles is just an example.

I do think you're going to confuse people by using "partition" to talk about browser profiles. Other web specs are tending toward "partitions" being aligned with top-level sites, and toward letting [some sorts of data](https://developer.mozilla.org/en-US/docs/Web/API/Document/requestStorageAccess) cross partition boundaries.

Reply:
Yes, I also agree that partition isn't a good fit; I used it above just to avoid introducing a new term.

the more I think about this, the more that I think the gecko-internal term of "user context" makes some sense; in that sense a single user agent (in the pre-infra sense) can be divided into multiple "user contexts" which represent a user-controlled partitioning of the browser state.

Reply:
FYI, we're also defining "context" in https://w3ctag.github.io/privacy-principles/#context, but that's not exported, and I don't think it'll be used from enough places that you need to worry about using a different meaning in Web Driver.